### PR TITLE
Refactor serializers (and service factory)

### PR DESCRIPTION
**Pull request by aluu317**
This PR:
- Moves the logic for the below from `service_factory` into `serializers`:
```
def create_request_data_model() -> Type[DataBase]:
    # return the data model class describing the request message

def create_rpc_json() -> Dict:
    # return json snippet for the service definition of this RPC
```
- Adds a few more tests for `service_factory` test coverage

Closes https://github.com/caikit/caikit/issues/53

### CODE REVIEW

File: caikit/runtime/service_factory.py [hunk 1]
```diff
@@ -216,18 +202,13 @@ def get_service_package(
...
             task_rpc_list = [
                 rpc for rpc in task_rpc_list if rpc.return_type is not None
             ]
 
-            request_data_models = cls._create_request_message_types(
-                task_rpc_list, package_name
-            )
+            request_data_models = []
+            for rpc in task_rpc_list:
+                request_data_models.append(rpc.create_request_data_model(package_name))
```
Comment (joerunde) [inline caikit/runtime/service_factory.py:211]:
```
request_data_models = [
    rpc.create_request_data_model(package_name) for rpc in task_rpc_list
]
```
?

File: caikit/runtime/service_factory.py [hunk 2]
```diff
@@ -238,7 +219,10 @@ def get_service_package(
                 # We need the message class that data model serializes to
                 setattr(client_module, dm_class.__name__, type(dm_class().to_proto()))
 
-            service_json = cls._create_service_json(task_rpc_list, package_name)
+            rpc_jsons = []
+            for rpc in task_rpc_list:
+                rpc_jsons.append(rpc.create_rpc_json(package_name))
```
Comment (joerunde) [inline caikit/runtime/service_factory.py:224]: could probably replace with a list comprehension as well

File: caikit/runtime/service_factory.py [hunk 3]
```diff
@@ -323,71 +307,6 @@ def _get_and_filter_modules(
...
-    @staticmethod
-    def _create_service_json(
-        rpcs_list: List[RPCSerializerBase], package_name: str
-    ) -> Dict:
-        """Make a json service def out of some rpc defs"""
-        rpc_jsons = []
-        for task in rpcs_list:
-            # The return type should be a "data model" object.
-            # We can take advantage of the fact that all of these should contain a private
-            # `proto_class` field ...and use that to yoink the fully qualified name of the
```
Comment (joerunde) [inline caikit/runtime/service_factory.py:377]: hah, that was a pretty old comment

File: caikit/runtime/service_factory.py (hunk 2 above)
Comment (joerunde) [reply caikit/runtime/service_factory.py:224]: oh man, just re-reading this code and the logic for creating the `package_name` is gnarly lol.

Gotta think about that one more, maybe when we introduce `@task`s they should come with a package name?

File: caikit/runtime/service_generation/create_service.py [hunk 4]
```diff
@@ -55,6 +55,7 @@ def create_inference_rpcs(modules: List[Type[ModuleBase]]) -> List[RPCSerializer
 
     rpcs = []
     # Inference specific logic:
+    # Remove non-primitive modules (including modules that return None type)
```
Comment (joerunde) [inline caikit/runtime/service_generation/create_service.py:58]: cannot wait to not have to do this 🙃

File: caikit/runtime/service_generation/serializers.py [hunk 5]
```diff
@@ -50,6 +61,52 @@ def module_list(self) -> List[Type[ModuleBase]]:
     def request(self) -> "_RequestMessage":
         """Return the internal representation of the request message type for this RPC"""
 
+    def create_request_data_model(self, package_name: str) -> Type[DataBase]:
+        """Dynamically create data model for this class's input RPC"""
```
Comment (joerunde) [inline caikit/runtime/service_generation/serializers.py:65]: 🤔 
`Dynamically create data model for this RPC's request message`?

There might also be a fun edge case here where the class is created multiple times if this method is called multiple times. I don't know if that would be a problem or not, might be fun to at least throw in a test that makes sure nothing blows up, if one doesn't already exist

Comment (joerunde) [reply caikit/runtime/service_generation/serializers.py:65]: (I'm guessing it's okay to do since this class would be stored in some `locals` that will be different on each invocation)

File: tests/runtime/test_service_factory.py [hunk 6]
```diff
@@ -63,6 +63,18 @@ def compiled_training_service() -> ServicePackage:
     )
 
 
+def test_service_package_raises_for_compiled_training_management():
+    with pytest.raises(CaikitRuntimeException) as e:
+        ServicePackageFactory.get_service_package(
+            ServicePackageFactory.ServiceType.TRAINING_MANAGEMENT,
+            ServicePackageFactory.ServiceSource.COMPILED,
+        )
+    assert (
```
Comment (joerunde) [inline tests/runtime/test_service_factory.py:72]: something cool I found the other day is that `pytest.raises` has a `match` keyword that takes regexes to match on the exception message so that you don't have to do this!

```
with pytest.raises(CaikitRuntimeException, match="Not allowed to get Training Management services from compiled packages"):
```

File: tests/runtime/test_service_factory.py [hunk 7]
```diff
@@ -313,6 +323,5 @@ def test_get_and_filter_modules_respects_included_task_types_and_excluded_module
     ) as cfg:
 
         clean_modules = ServicePackageFactory._get_and_filter_modules(cfg, "sample_lib")
-        assert len(clean_modules) == 4
```
Comment (joerunde) [inline tests/runtime/test_service_factory.py:316]: nice, good to see tests more robust against test fixture changes 🌶️

**Comment by joerunde** [comment]
Looks great!

We should get some better names in for these, maybe a starting point is to rename:
`serializers.py` -> `rpcs.py`
`RPCSerializerBase` -> `CaikitRPC`
?

File: caikit/runtime/service_factory.py (hunk 2 above)
Comment (aluu317) [reply caikit/runtime/service_factory.py:224]: Now that you mention. I think this logic is redundant:
```
ai_domain_name = snake_to_upper_camel(lib.replace("caikit_", ""))
```
if our `lib` is sample_lib. This might have been more for our internal use case.

I think we need to revisit this when `task decorator` is introduced

**Review by joerunde — APPROVED** [review]
j'approuve, these improvements are 🌭🌭🌭
